File: leetcode/add-two-numbers/AddTwoNumbers.cpp
#include <stddef.h> // NULL
#include <stdio.h>  // printf

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    // cpp allow constructor in struct
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode head(0); // extra head node
        ListNode *curr = &head; // pointer to move next(shoud use & to get the address of head!!)
        int sum = 0;
        while (l1 != NULL && l2 != NULL) { // NULL, not null
            sum = sum / 10 + l1->val + l2->val; // which is of pointer type ‘ListNode*’ (maybe you meant to use ‘->’ ?)
            curr->next = new ListNode(sum % 10);
            curr = curr->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        // handle the rest
        ListNode *rest = l1 != NULL ? l1 : l2;
        while (rest != NULL) {
            sum = sum / 10 + rest->val;
            curr->next = new ListNode(sum % 10);
            curr = curr->next;
            rest = rest->next;
        }
        // handle the carry
        printf("carry: %d\n", sum / 10);
        if (sum / 10 != 0) {
            curr->next = new ListNode(1);
        }
        return head.next;
    }

    ListNode* addTwoNumbers2(ListNode* l1, ListNode* l2) {
        ListNode *head = NULL;
        ListNode **curr = &head; // use secondary pointer
        int sum = 0;
        while (l1 != NULL || l2 != NULL) { // NULL, not null
            sum /= 10; // add the carry first
            if (l1 != NULL) {
                sum += l1->val;
                l1 = l1->next;
            }
            if (l2 != NULL) {
                sum += l2->val;
                l2 = l2->next;
            }
            *curr = new ListNode(sum % 10);
            curr = &((*curr)->next);
        }
        // check if has carry after all finished
        if (sum / 10 != 0) {
            *curr = new ListNode(1);
        }
        return head;
    }

    ListNode* createList(int arr[], int len) {
        // int len = sizeof(arr) / sizeof(int) + 1; // wrong, arrary should specify length
        printf("length: %d\n", len);
        ListNode head(0);
        ListNode *curr = &head;
        for (int i = 0; i < len; i++) {
            curr->next = new ListNode(arr[i]);
            curr = curr->next;
        }
        return head.next;
    }
    void printList(ListNode* list) {
        while(list != NULL) {
            printf("%d->", list->val);
            list = list->next;
        }
        printf("null\n");
    }
};
int main() {
    Solution sol;
    int arr1[] = {1, 2, 3};
    int arr2[] = {4, 5, 7};
    ListNode* l1 = sol.createList(arr1, 3);
    ListNode* l2 = sol.createList(arr2, 3);
    ListNode* l3 = sol.addTwoNumbers(l1, l2);
    sol.printList(l1);
    sol.printList(l2);
    sol.printList(l3);
    return 0;
}

File: leetcode/add-two-numbers/AddTwoNumbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:

    # Add Two Numbers.
    # You are given two linked lists representing two non-negative numbers.
    # The digits are stored in reverse order and each of their nodes contain a single digit.
    # Add the two numbers and return it as a linked list.
    # Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
    # Output: 7 -> 0 -> 8
    def addTwoNumbers(self, l1, l2):
        head = ListNode(0) # needn't new to create an instance
        curr = head
        sum = 0
        while l1 != None and l2 != None:
            sum = l1.val + l2.val + sum / 10
            curr.next = ListNode(sum % 10)
            l1 = l1.next
            l2 = l2.next
            curr = curr.next
        # handle the rest condition
        if l1 == None:
            rest = l2
        else:
            rest = l1
        # no carry, set the rest link as next
        if sum / 10 == 0:
            curr.next = rest
            return head.next
        # has carry
        while rest != None:
            sum = rest.val + sum / 10
            curr.next = ListNode(sum % 10)
            rest = rest.next
            curr = curr.next
        if sum / 10 == 1:
            curr.next = ListNode(1)
        return head.next

    def createList(self, arr):
        head = ListNode(0)
        curr = head
        for val in arr:
            curr.next = ListNode(val)
            curr = curr.next
        return head.next

    def printList(self, ll):
        while ll != None:
            print ll.val, " -> ", # put `,` in the end(python 2.x) to make sure it will not print '\n'
            ll = ll.next
        print "None"

if __name__ == '__main__':
    sol = Solution()
    arr1, arr2 = [2, 4, 3], [5, 6, 4]
    l1, l2 = sol.createList(arr1), sol.createList(arr2)
    sol.printList(l1)
    sol.printList(l2)
    result = sol.addTwoNumbers(l1, l2)
    sol.printList(result)
